### PR TITLE
checksum: simplify, use only sha256

### PR DESCRIPTION
**Pull request by iMichka**
We use only one sha type right now.

Needed for https://github.com/Homebrew/brew/pull/10186

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

### CODE REVIEW

File: Library/Homebrew/cmd/fetch.rb [hunk 1]
```diff
@@ -183,7 +183,7 @@ def fetch_fetchable(f, args:)
     return unless download.file?
 
     puts "Downloaded to: #{download}" unless already_fetched
-    puts Checksum::TYPES.map { |t| "#{t.to_s.upcase}: #{download.send(t)}" }
+    puts "#{:sha256.to_s.upcase}: #{download.send(:sha256)}"
```
Comment (reitermarkus) [inline Library/Homebrew/cmd/fetch.rb:186]:
```suggestion
    puts "SHA256: #{download.sha256}"
```

File: Library/Homebrew/extend/pathname.rb [hunk 2]
```diff
@@ -303,7 +303,7 @@ def sha256
   def verify_checksum(expected)
     raise ChecksumMissingError if expected.blank?
 
-    actual = Checksum.new(expected.hash_type, send(expected.hash_type).downcase)
+    actual = Checksum.new(send(:sha256).downcase)
```
Comment (reitermarkus) [inline Library/Homebrew/extend/pathname.rb:306]:
```suggestion
    actual = Checksum.new(sha256.downcase)
```

Comment (reitermarkus) [reply Library/Homebrew/extend/pathname.rb:306]: May be best to move the `.downcase` straight into `sha256`.

File: Library/Homebrew/formula.rb [hunk 3]
```diff
@@ -2418,9 +2418,7 @@ def mirror(val)
     # tell you the currently valid value.
     #
     # <pre>sha256 "2a2ba417eebaadcb4418ee7b12fe2998f26d6e6f7fda7983412ff66a741ab6f7"</pre>
-    Checksum::TYPES.each do |type|
-      define_method(type) { |val| stable.send(type, val) }
-    end
+    define_method(:sha256) { |val| stable.send(:sha256, val) }
```
Comment (reitermarkus) [inline Library/Homebrew/formula.rb:2421]:
```suggestion
    define_method(:sha256) { |val| stable.sha256 val }
```

File: Library/Homebrew/resource.rb [hunk 4]
```diff
@@ -162,9 +162,7 @@ def verify_download_integrity(fn)
     EOS
   end
 
-  Checksum::TYPES.each do |type|
-    define_method(type) { |val| @checksum = Checksum.new(type, val) }
-  end
+  define_method(:sha256) { |val| @checksum = Checksum.new(val) }
```
Comment (reitermarkus) [inline Library/Homebrew/resource.rb:165]:
```suggestion
  def sha256(val) 
    @checksum = Checksum.new(val)
  end
```

File: Library/Homebrew/software_spec.rb [hunk 5]
```diff
@@ -411,11 +411,9 @@ def tag?(tag)
 
   # Checksum methods in the DSL's bottle block optionally take
   # a Hash, which indicates the platform the checksum applies on.
-  Checksum::TYPES.each do |cksum|
-    define_method(cksum) do |val|
-      digest, tag = val.shift
-      collector[tag] = Checksum.new(cksum, digest)
-    end
+  define_method(:sha256) do |val|
```
Comment (reitermarkus) [inline Library/Homebrew/software_spec.rb:414]:
```suggestion
  def sha256(val)
```

File: Library/Homebrew/formula.rb [hunk 6]
```diff
@@ -2418,9 +2418,7 @@ def mirror(val)
     # tell you the currently valid value.
     #
     # <pre>sha256 "2a2ba417eebaadcb4418ee7b12fe2998f26d6e6f7fda7983412ff66a741ab6f7"</pre>
-    Checksum::TYPES.each do |type|
-      define_method(type) { |val| stable.send(type, val) }
-    end
+    define_method(:sha256) { |val| stable.sha256 val }
```
Comment (reitermarkus) [inline Library/Homebrew/formula.rb:2421]:
```suggestion
    delegate sha256: :stable
```

Comment (iMichka) [reply Library/Homebrew/formula.rb]: This gives the following error:
 `Error: unknown keyword: sha256`

**Comment by iMichka** [comment]
There are still a few places that call `hash_type` and I need to check these. I pushed a few more changes.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, nice cleanup @iMichka! https://github.com/Homebrew/brew/blob/db0b69ac6ab1fc6dc4c8ad466a5de9105c1b54bf/Library/Homebrew/exceptions.rb#L627-L643 still needs to be updated, I think, but once that's done and CI is 🟢 I'm 👍🏻 to 🚢

**Comment by iMichka** [comment]
All green for me, except the `delegate` change that I am not able to make work.

File: Library/Homebrew/formula.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:2421]: @iMichka Try to add `:sha256,` to here instead:

https://github.com/Homebrew/brew/blob/a65ba2fbfedca0ace75af69666848258916e677e/Library/Homebrew/formula.rb#L341-L350

Comment (reitermarkus) [reply Library/Homebrew/formula.rb:2421]: Missing `extend Forwardable` maybe?

Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:2421]: It's using `delegate` elsewhere and I actually think it should be forwarded to the current spec anyway.

Comment (iMichka) [reply Library/Homebrew/formula.rb:2421]: This method definition is in the `class << self` block. I see that there is no other delegation there, so I guess this is not possible.

Comment (SeekingMeaning) [reply Library/Homebrew/formula.rb:2421]: `class << self` possibly needs `extend Forwardable`

Comment (SeekingMeaning) [reply Library/Homebrew/formula.rb:2421]: > It's using `delegate` elsewhere and I actually think it should be forwarded to the current spec anyway.

No, I think this is correct and should be using `stable`

**Comment by SeekingMeaning** [comment]
I would recommend this:

```rb
    def sha256(val)
      stable.sha256(val)
    end
```

Or otherwise using `delegate`: https://github.com/SeekingMeaning/brew/commit/9faf25527

File: Library/Homebrew/formula.rb (hunk 6 above)
Comment (iMichka) [reply Library/Homebrew/formula.rb]: I used the same syntax as def mirror above. Works for me.
We can have a second PR to make it forwardable and use delegate if you want.

**Comment by SeekingMeaning** [comment]
Probably should try to add as few `:sha256` symbols as possible; we can delete the `hash_type` variable in `bump-formula-pr`: https://github.com/SeekingMeaning/brew/commit/7c248d26a

Currently looking into `def checksums` in `software_spec.rb`—we might be able to return an array instead of a hash that only has `:sha256` as a key

**Comment by SeekingMeaning** [comment]
Question: Are there any plans for SHA-512?

**Comment by iMichka** [comment]
> bump-formula-pr

Done

**Comment by SeekingMeaning** [comment]
(You missed the `hash_type.blank?` &rarr; `old_hash.blank?` change)

**Review by SeekingMeaning — APPROVED** [review]
I would also recommend https://github.com/SeekingMeaning/brew/commit/c01b1cc80 but I'm not familiar with all uses of `BottleSpecification#checksums` so 👍 from me

**Comment by iMichka** [comment]
The checksums are coming from https://github.com/Homebrew/brew/pull/10247/files#diff-bb40db554b6955f972e80737c5883dafef80ffe19b09e0ad1ceeea1cb776f694L425-L439 for this part of the code. This should be fine. This is a part of code I will rework with my follow up PRs as the logic is going to change there.

File: Library/Homebrew/software_spec.rb [hunk 7]
```diff
@@ -432,8 +430,8 @@ def checksums
     checksums = {}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:430]:
```suggestion
    sha256s = []
```

File: Library/Homebrew/software_spec.rb [hunk 8]
```diff
@@ -432,8 +430,8 @@ def checksums
...
     tags.reverse_each do |tag|
       checksum = collector[tag]
-      checksums[checksum.hash_type] ||= []
-      checksums[checksum.hash_type] << { checksum => tag }
+      checksums[:sha256] ||= []
+      checksums[:sha256] << { checksum => tag }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:434]:
```suggestion
      sha256s << { checksum => tag }
```

File: Library/Homebrew/software_spec.rb [hunk 9]
```diff
@@ -432,8 +430,8 @@ def checksums
...
     end
     checksums
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:436]:
```suggestion
   { sha256: sha256s }
```

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks for the cleanup @iMichka!

**Pull request merged**